src/swaps/origins/OriginsSwapProvider.js

```diff
@@ -0,0 +1,346 @@
+import BN from 'bignumber.js'
+import { v4 as uuidv4 } from 'uuid'
+import * as ethers from 'ethers'
+
+import { chains, currencyToUnit, unitToCurrency } from '@liquality/cryptoassets'
+import cryptoassets from '@/utils/cryptoassets'
+import { isERC20 } from '../../utils/asset'
+import { prettyBalance } from '../../utils/coinFormatter'
+import { ChainNetworks } from '@/utils/networks'
+import { withInterval, withLock } from '../../store/actions/performNextAction/utils'
+import store from '../../store'
+import { SwapProvider } from '../SwapProvider'
+import ERC20 from '@uniswap/v2-core/build/ERC20.json'
+
+import controllerABI from './abi/controllerAbi.json'
+import presaleABI from './abi/presale.json'
+
+class OriginsSwapProvider extends SwapProvider {
+  constructor (config) {
+    super(config)
+    this._apiCache = {} // chainId to RPC provider
+  }
+
+  async getSupportedPairs () {
+    return []
+  }
+
+  // returns rates between tokens
+  async getQuote ({ network, from, to, amount }) {
+    const fromInfo = cryptoassets[from]
+    const toInfo = cryptoassets[to]
+
+    // only RSK network swaps
+    if (fromInfo.chain !== 'rsk' || toInfo.chain !== 'rsk' || amount <= 0) return null
+
+    // this provider supports only SOV->ZERO.
+    if (from !== 'SOV' || to !== 'ZERO') return null
+
+    const isClosed = await this.checkSaleClosed(network)
+    if (isClosed) return null
+
+    // calculate rates
+    const rate = await this.getDepositRate(network)
+
+    const fromAmountInUnit = currencyToUnit(fromInfo, BN(amount)).toFixed()
+    const toAmountInUnit = BN(fromAmountInUnit).times(rate)
+
+    return {
+      from,
+      to,
+      fromAmount: fromAmountInUnit,
+      toAmount: toAmountInUnit
+    }
+  }
+
+  async checkSaleClosed (network) {
+    const erc20 = new ethers.Contract(this.config.presaleAddress, presaleABI, this._getApi(network, 'ZERO'))
+    const isClosed = await erc20.isClosed()
+    return isClosed
+  }
+
+  async getDepositRate (network) {
+    const erc20 = new ethers.Contract(this.config.presaleAddress, presaleABI, this._getApi(network, 'ZERO'))
+
+    const PPM = (await erc20.PPM()).toString()
+    const exchangeRate = (await erc20.exchangeRate()).toString()
+    return BN(exchangeRate).dividedBy(BN(PPM)).toNumber()
```

Comment:
Using `toNumber()` might cause precision loss due to scientific notation.

Reply:
I changed it to `toString()`. thanks!